File: fastsort/fastsort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void quicksort(int A[10], int p, int r){
	if(p<r){
	int q;
	q = my_PARTITION(A, p, r);
	quicksort(A, p, q-1);
	quicksort(A, q+1, r);
	}
}

int my_PARTITION(int A[8], int p, int r){
	printf("We steped in partition!\n");
	int x = A[r-1];
	int i = p-1;
	int m;
	int s;
	
	printf("x: %d\n", x);

	for(int j=p; j<r-1; j++){
		printf("aj: %d\n",A[j]);
		if(A[j]<=x){
		i=i+1;
		m=A[j], A[j]=A[i], A[i]=m;
		}
	printf("i:\t%d,j:\t%d\n", i,j);
	for(int k=0;k<r;k++)
		printf("%d", A[k]);
	printf("\n");
	
	}
	s=A[r-1], A[r-1]=A[i+1], A[i+1]=s;
        printf("================Partition fullfilled!===============\n");
	return i+1;
}


int main(){
	int arr[8] = {2,8,7,1,3,5,6,4};

	printf("===================Array initialized !==================\n");

	quicksort(arr, 0, 8);

	for(int i=0;i<8;i++)
		printf("%d\n", arr[i]);
	return 0;
	}

File: Sorting/fast_sort.py
#py3

def fast_sort(nums, start, end):
    #divide
    def partition(arr, l, r):
        #print(arr, r)
        key = arr[r]
        s = l - 1

        for i in range(l, r):
            if arr[i] <= key:
                s += 1
                arr[s], arr[i] = arr[i], arr[s]
        
        arr[s+1], arr[r] = arr[r], arr[s+1]
        return s+1
    
    if start < end:
        part = partition(nums, start, end)
        fast_sort(nums, start, part-1)
        fast_sort(nums, part+1, end)
    #return nums


def quicksort(arr, p, r):
        if p < r:
                q = my_PARTITION(arr, p, r)
                quicksort(arr, p, q-1)
                quicksort(arr, q+1, r)



def my_PARTITION(arr, p, r):
        #set `x` as metric
        x = arr[r]
        i = p-1

        for j in range(p, r):
                #if arr[j] smaller than metric, swap j to left, which is the region smaller than metric
                if arr[j] <= x:
                        i = i + 1
                        arr[i], arr[j] = arr[j], arr[i]
        #swap the metric with index of i, which is the greatest one
        arr[r], arr[i+1] = arr[i+1], arr[r]

        return i+1



def testcase():
    arr = [6, 3,5,4,1,7,9,2,6,4,8,9,1,3,5,7,9,2,3]
    #arr = [2,8,7,1,3,5,6,4]
    fast_sort(arr, 0, len(arr)-1)
    print(arr)


testcase()


